postgres-appliance/major_upgrade/inplace_upgrade.py

```diff
@@ -0,0 +1,727 @@
+#!/usr/bin/env python
+import json
+import logging
+import os
+import psutil
+import psycopg2
+import shlex
+import shutil
+import subprocess
+import sys
+import time
+import yaml
+
+from collections import defaultdict
+from threading import Thread
+from multiprocessing.pool import ThreadPool
+
+logger = logging.getLogger(__name__)
+
+
+def patch_wale_prefix(value, new_version):
+    from spilo_commons import is_valid_pg_version
+
+    if '/spilo/' in value and '/wal/' in value:  # path crafted in the configure_spilo.py?
+        basename, old_version = os.path.split(value.rstrip('/'))
+        if is_valid_pg_version(old_version) and old_version != new_version:
+            return os.path.join(basename, new_version)
+    return value
+
+
+def update_configs(new_version):
+    from spilo_commons import append_extentions, get_bin_dir, get_patroni_config, write_file, write_patroni_config
+
+    config = get_patroni_config()
+
+    config['postgresql']['bin_dir'] = get_bin_dir(new_version)
+
+    version = float(new_version)
+    shared_preload_libraries = config['postgresql'].get('parameters', {}).get('shared_preload_libraries')
+    if shared_preload_libraries is not None:
+        config['postgresql']['parameters']['shared_preload_libraries'] =\
+                append_extentions(shared_preload_libraries, version)
+
+    extwlist_extensions = config['postgresql'].get('parameters', {}).get('extwlist.extensions')
+    if extwlist_extensions is not None:
+        config['postgresql']['parameters']['extwlist.extensions'] =\
+                append_extentions(extwlist_extensions, version, True)
+
+    write_patroni_config(config, True)
+
+    # update wal-e/wal-g envdir files
+    restore_command = shlex.split(config['postgresql'].get('recovery_conf', {}).get('restore_command', ''))
+    if len(restore_command) > 4 and restore_command[0] == 'envdir':
+        envdir = restore_command[1]
+
+        try:
+            for name in os.listdir(envdir):
+                if len(name) > 12 and name.endswith('_PREFIX') and name[:5] in ('WALE_', 'WALG_'):
+                    name = os.path.join(envdir, name)
+                    try:
+                        with open(name) as f:
+                            value = f.read().strip()
+                        new_value = patch_wale_prefix(value, new_version)
+                        if new_value != value:
+                            write_file(new_value, name, True)
+                    except Exception as e:
+                        logger.error('Failed to process %s: %r', name, e)
+        except Exception:
+            pass
```

Comment:
what exactly is handled here ? inaccessible `envdir` ?

Reply:
Yes, non-existing or inaccessible.